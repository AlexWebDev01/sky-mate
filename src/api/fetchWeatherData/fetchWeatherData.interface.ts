import { WeatherConditions } from "../../shared/constants/clothesAlgorithm/clothesAlgorithm.interface";

export interface WeatherData {
  current: CurrentWeatherData;
  daily: DailyWeatherData[];
  hourly: HourlyWeatherData[];
  lat: number;
  lon: number;
  timezone: string;
  timezone_offset: number;
}

export interface CurrentWeatherData extends BaseWeatherData {
  sunrise: number;
  sunset: number;
  visibility: number;
  temp: number;
  feels_like: number;
}

export interface HourlyWeatherData extends BaseWeatherData {
  pop: number;
  wind_gust: number;
  visibility: number;
  temp: number;
  feels_like: number;
}

export interface DailyWeatherData extends BaseWeatherData {
  feels_like: { day: number; night: number; eve: number; morn: number };
  moon_phase: number;
  moonrise: number;
  moonset: number;
  pop: number;
  rain: number;
  summary: string;
  sunrise: number;
  sunset: number;
  temp: {
    day: number;
    eve: number;
    max: number;
    min: number;
    morn: number;
    night: number;
  };
  wind_gust: number;
}

interface BaseWeatherData {
  clouds: number;
  dew_point: number;
  dt: number;
  humidity: number;
  pressure: number;
  uvi: number;
  weather: Weather[];
  wind_deg: number;
  wind_speed: number;
}

export interface Weather {
  description: string;
  icon: string;
  id: number;
  main: WeatherConditions;
}

export interface Coordinates {
  lat: number;
  lon: number;
}

export enum Units {
  standard = "standard",
  metric = "metric",
  imperial = "imperial",
}
