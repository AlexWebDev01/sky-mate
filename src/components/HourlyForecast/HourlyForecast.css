@import url("../../shared/assets/fonts/fonts.css");
.hourly-forecast {
  position: relative;
  font-family: Oddval;
}
.hourly-forecast .slider-container {
  display: flex;
  flex-direction: row;
  overflow: hidden;
}
.hourly-forecast .slider-container .slider {
  display: flex;
  transition: transform 0.3s ease;
}
.hourly-forecast .slider-container .slider .day {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-around;
  width: 55px;
  height: 65px;
}
.hourly-forecast .slider-container .slider .day .line {
  width: 100%;
  height: 1px;
  background-color: var(--white);
}
.hourly-forecast .right-arrow,
.hourly-forecast .left-arrow {
  position: absolute;
  display: flex;
  height: auto;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  background-color: transparent;
  border: none;
  padding: 0;
}
.hourly-forecast .left-arrow {
  left: -20px;
  top: 22px;
}
.hourly-forecast .right-arrow {
  top: 21px;
  right: -20px;
}/*# sourceMappingURL=HourlyForecast.css.map */