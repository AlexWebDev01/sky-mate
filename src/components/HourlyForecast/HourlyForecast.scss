@import url('../../shared/assets/fonts/fonts.css');

.hourly-forecast {
  position: relative;
  font-family: Oddval;

  .slider-container {
    display: flex;
    flex-direction: row;
    overflow: hidden;

    .slider {
      display: flex;
      transition: transform 0.3s ease;

      .day {
        display: flex;
        flex-direction: column;
        align-items: center;
        justify-content: space-around;
        width: 55px;
        height: 65px;

        .line {
          width: 100%;
          height: 1px;
          background-color: var(--white);
        }
      }
    }
  }

  .right-arrow,
  .left-arrow {
    position: absolute;
    display: flex;
    height: auto;
    justify-content: center;
    align-items: center;
    cursor: pointer;
    background-color: transparent;
    border: none;
    padding: 0;
  }

  .left-arrow {
    left: -20px;
    top: 22px;
  }

  .right-arrow {
    top: 21px;
    right: -20px;
  }
}
