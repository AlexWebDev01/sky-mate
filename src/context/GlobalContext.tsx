import { useState, createContext, useContext, FunctionComponent } from "react";
import {
  Coordinates,
  GlobalContext,
  GlobalProviderProps,
} from "./GlobalContext.interface";

const GlobalContext = createContext<GlobalContext | null>(null);

export const useGlobalContext = () => {
  const context = useContext(GlobalContext);

  if (!context) {
    throw new Error("useWords must be used within a WordsProvider");
  }

  return context;
};

export const GlobalProvider: FunctionComponent<GlobalProviderProps> = ({
  children,
}) => {
  const [coordinates, setCoordinates] = useState<Coordinates | null>(null);
  const [weatherData, setWeatherData] = useState<object | null>(null);
  const [location, setLocation] = useState<string | null>(null);
  const [pageStyle, setPageStyle] = useState<string>("");

  const sharedState: GlobalContext = {
    coordinates,
    setCoordinates,
    weatherData,
    setWeatherData,
    location,
    setLocation,
    pageStyle,
    setPageStyle,
  };

  return (
    <GlobalContext.Provider value={sharedState}>
      {children}
    </GlobalContext.Provider>
  );
};
