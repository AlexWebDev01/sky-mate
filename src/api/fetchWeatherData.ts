import { Coordinates } from "../context/GlobalContext.interface";

export const fetchWeatherData = async (
  coordinates: Coordinates,
  units: string
) => {
  console.log("COORDINATES:", coordinates);

  const response = await fetch(
    `${import.meta.env.VITE_WEATHER_APP_API_URL}/onecall?lat=${
      coordinates.lat
    }&lon=${coordinates.lon}&units=${units}&exclude=minutely&appid=${
      import.meta.env.VITE_WEATHER_APP_API_KEY
    }`
  );

  const data = await response.json();

  return data;
};
