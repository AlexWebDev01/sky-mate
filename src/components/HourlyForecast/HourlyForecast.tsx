import { useState } from 'react';
import { useGlobalContext } from '../../context/GlobalContext';

import HourlyForecastItem from './components/HourlyForecastItem';

import { HourlyWeatherData } from '../../context/GlobalContext.interface';

import './HourlyForecast.css';

interface Props {
  expanded: boolean;
}

const HourlyForecast = ({ expanded }: Props) => {
  const [currentPosition, setCurrentPosition] = useState(0);

  const { state } = useGlobalContext();
  const { weatherData } = state;

  const sliderItems: HourlyWeatherData[] = weatherData.hourly.slice(1, 14);
  const itemWidth = 55;
  const isLeftArrowDisabled = currentPosition <= 0;
  const isRightArrowDisabled = currentPosition >= 220;

  const slideLeft = () => {
    const newPosition = currentPosition - itemWidth;
    setCurrentPosition(newPosition);
  };

  const slideRight = () => {
    const newPosition = currentPosition + itemWidth;
    setCurrentPosition(newPosition);
  };

  return (
    <div className={expanded ? `hourly-forecast` : 'hide'}>
      <button
        className='left-arrow'
        onClick={slideLeft}
        disabled={isLeftArrowDisabled}
      >
        {/* TODO: import svg icons to React projects */}
        <svg
          width='11'
          height='19'
          viewBox='0 0 11 19'
          fill='none'
          xmlns='http://www.w3.org/2000/svg'
        >
          <path
            d='M8.82028 18.624L0.278817 10.2018C0.177435 10.1016 0.105453 9.99296 0.0628727 9.87599C0.0202921 9.75902 -0.000660041 9.63369 1.58411e-05 9.5C1.58411e-05 9.36631 0.020968 9.24098 0.0628727 9.12401C0.104777 9.00704 0.176759 8.89842 0.278817 8.79815L8.82028 0.350923C9.05684 0.116974 9.35254 0 9.70738 0C10.0622 0 10.3664 0.12533 10.6198 0.375989C10.8733 0.626649 11 0.919085 11 1.2533C11 1.58751 10.8733 1.87995 10.6198 2.13061L3.16821 9.5L10.6198 16.8694C10.8564 17.1033 10.9747 17.3918 10.9747 17.7347C10.9747 18.0776 10.8479 18.374 10.5945 18.624C10.341 18.8747 10.0453 19 9.70738 19C9.36943 19 9.07374 18.8747 8.82028 18.624Z'
            fill='white'
          />
        </svg>
      </button>
      <div className='slider-container'>
        <div
          id='slider'
          className='slider'
          style={{ transform: `translateX(-${currentPosition}px)` }}
        >
          {sliderItems.map(
            (item: HourlyWeatherData, timezoneOffset: number) => (
              <HourlyForecastItem
                item={item}
                timezoneOffset={timezoneOffset}
                key={item.dt}
              />
            )
          )}
        </div>
      </div>
      <button
        className='right-arrow'
        onClick={slideRight}
        disabled={isRightArrowDisabled}
      >
        <svg
          width='11'
          height='19'
          viewBox='0 0 11 19'
          fill='none'
          xmlns='http://www.w3.org/2000/svg'
        >
          <path
            d='M2.17972 18.624L10.7212 10.2018C10.8226 10.1016 10.8945 9.99296 10.9371 9.87599C10.9797 9.75902 11.0007 9.63369 11 9.5C11 9.36631 10.979 9.24098 10.9371 9.12401C10.8952 9.00704 10.8232 8.89842 10.7212 8.79815L2.17972 0.350923C1.94316 0.116974 1.64746 0 1.29262 0C0.937786 0 0.633639 0.12533 0.380184 0.375989C0.126728 0.626649 0 0.919085 0 1.2533C0 1.58751 0.126728 1.87995 0.380184 2.13061L7.83179 9.5L0.380184 16.8694C0.143625 17.1033 0.0253448 17.3918 0.0253448 17.7347C0.0253448 18.0776 0.152073 18.374 0.405529 18.624C0.658985 18.8747 0.954683 19 1.29262 19C1.63057 19 1.92626 18.8747 2.17972 18.624Z'
            fill='white'
          />
        </svg>
      </button>
    </div>
  );
};

export default HourlyForecast;
