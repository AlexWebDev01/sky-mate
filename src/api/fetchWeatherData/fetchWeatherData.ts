import { Units, Coordinates, WeatherData } from "./fetchWeatherData.interface";

export const fetchWeatherData = async (
  coordinates: Coordinates,
  units: Units
): Promise<WeatherData> => {
  const response = await fetch(
    `${import.meta.env.VITE_WEATHER_APP_API_URL}/onecall?lat=${
      coordinates.lat
    }&lon=${coordinates.lon}&units=${units}&exclude=minutely&appid=${
      import.meta.env.VITE_WEATHER_APP_API_KEY
    }`
  );

  const data: WeatherData = await response.json();

  return data;
};
