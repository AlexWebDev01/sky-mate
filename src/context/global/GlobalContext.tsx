import {
  useState,
  createContext,
  useContext,
  FunctionComponent,
  useCallback,
  useMemo,
} from 'react';
import {
  GlobalContext as GlobalContextInterface,
  GlobalProviderProps,
  GlobalState,
} from './GlobalContext.interface';

import { fetchLocationData } from '@api/fetchLocationData/fetchLocationData';
import { fetchUserLocatioByIP } from '@api/fetchUserLocationByIP/fetchUserLocationByIP';
import { separateCoordinates } from '@shared/helpers/date/date';
import { fetchWeatherData } from '@api/fetchWeatherData/fetchWeatherData';
import { calculateClothesAdvice } from '@shared/helpers/clothesAlgorithm';

import { WeatherConditions } from '@shared/constants/clothesAlgorithm/clothesAlgorithm.interface';
import { LocationDataByIP } from '@api/fetchUserLocationByIP/fetchUserLocationByIP.interface';
import { LocationData } from '@api/fetchLocationData/fetchLocationData.interface';
import {
  Units,
  Coordinates,
} from '@api/fetchWeatherData/fetchWeatherData.interface';

const GlobalContext = createContext<GlobalContextInterface | null>(null);

export const useGlobalContext = () => {
  const context = useContext(GlobalContext);

  if (!context) {
    throw new Error('Global context should be used in GlobalProvider!');
  }

  return context;
};

export const GlobalProvider: FunctionComponent<GlobalProviderProps> = ({
  children,
}) => {
  const [state, setState] = useState<GlobalState>({
    isLoading: true,
    coordinates: null,
    location: null,
    weatherData: null,
    pageStyle: '',
    clothesAdvice: null,
  });
  const [error, setError] = useState<string>('');

  const getWeatherData = useCallback(async (coordinates: Coordinates) => {
    const data = await fetchWeatherData(coordinates, Units.metric);
    const weatherCondition =
      data.daily[0].weather[0].main.toLowerCase() as WeatherConditions;
    const averageTemp = data.daily[0].temp.day;
    const clothesAdvice = calculateClothesAdvice(weatherCondition, averageTemp);
    return {
      weatherData: data,
      clothesAdvice,
      pageStyle: weatherCondition,
    };
  }, []);

  const fetchData = useCallback(async () => {
    console.group('FETCH DATA');
    console.info('Started initial data fetching');

    setState((prevState) => ({ ...prevState, isLoading: true }));

    try {
      const locationData: LocationDataByIP = await fetchUserLocatioByIP();
      const coordinates = separateCoordinates(locationData.loc);

      const { weatherData, clothesAdvice, pageStyle } =
        await getWeatherData(coordinates);
      setState((prevState) => ({
        ...prevState,
        weatherData,
        clothesAdvice,
        pageStyle,
        coordinates,
        location: locationData.city,
        isLoading: false,
      }));

      console.info('Fetched & calculated data: ', {
        weatherData,
        clothesAdvice,
        pageStyle,
        city: locationData.city,
      });
      console.info('Finished initial data fetching');
      console.groupEnd();
    } catch (error) {
      if (error instanceof Error) {
        setError('Oops, unexpected error!');
      }

      console.error('Failed to fetch initial data:', error);
    }
  }, [getWeatherData]);

  const handleSearch = useCallback(
    async (location: string) => {
      try {
        console.group('HANDLE SEARCH');
        console.info('Started fetching data for location:', location);
        const locationData: LocationData = await fetchLocationData(location);
        const { name, lat, lon } = locationData;

        const { weatherData, clothesAdvice, pageStyle } = await getWeatherData({
          lat,
          lon,
        });

        setState((prevState) => ({
          ...prevState,
          coordinates: { lat, lon },
          location: name,
          weatherData,
          clothesAdvice,
          pageStyle,
        }));
        console.info('Fetched & calculated data: ', {
          weatherData,
          clothesAdvice,
          pageStyle,
          city: name,
        });
        console.info('Finished fetching data for location:', location);
      } catch (error) {
        if (error instanceof Error && error.message === 'Location not found') {
          setError('Location not found');
        } else if (error instanceof Error) {
          setError('Oops, unexpected error!');
        }

        console.error('Error fetching data for location:', error);
      }
    },
    [getWeatherData]
  );

  const sharedState = useMemo<GlobalContextInterface>(
    () => ({
      state,
      error,
      setError,
      fetchData,
      handleSearch,
    }),
    [state, error, setError, fetchData, handleSearch]
  );

  return (
    <GlobalContext.Provider value={sharedState}>
      {children}
    </GlobalContext.Provider>
  );
};
