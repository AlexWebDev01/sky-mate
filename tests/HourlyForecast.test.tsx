import '@testing-library/jest-dom';
import { render, screen, fireEvent } from '@testing-library/react';
import HourlyForecast from '@components/HourlyForecast/HourlyForecast';
import { useGlobalContext } from '@context/global/GlobalContext';
import { WeatherData } from '@api/fetchWeatherData/fetchWeatherData.interface';
import {
  ClothesStyle,
  TemperatureRangeNames,
  WeatherConditions,
} from '@shared/constants/clothesAlgorithm/clothesAlgorithm.interface';

jest.mock('@shared/assets/icons/left-arrow.svg', () => {
  return () => <svg data-testid='mocked-left-arrow'></svg>;
});

jest.mock('@shared/assets/icons/right-arrow.svg', () => {
  return () => <svg data-testid='mocked-right-arrow'></svg>;
});

jest.mock('@context/global/GlobalContext', () => ({
  useGlobalContext: jest.fn(),
}));

const mockUseGlobalContext = useGlobalContext as jest.MockedFunction<
  typeof useGlobalContext
>;

describe('HourlyForecast Component', () => {
  const mockWeatherData: WeatherData = {
    current: {
      clouds: 75,
      dew_point: 10,
      dt: 1625235600,
      humidity: 65,
      pressure: 1013,
      uvi: 0.5,
      weather: [
        {
          id: 500,
          main: WeatherConditions.rain,
          description: 'light rain',
          icon: '10d',
        },
      ],
      wind_deg: 270,
      wind_speed: 5.1,
      sunrise: 1625217600,
      sunset: 1625268000,
      visibility: 10000,
      temp: 20,
      feels_like: 20,
    },
    daily: [],
    hourly: Array.from({ length: 14 }, (_, i) => ({
      dt: 1625235600 + i * 3600,
      temp: 20 + i,
      clouds: 75,
      dew_point: 10,
      humidity: 65,
      pressure: 1013,
      uvi: 0.5,
      weather: [
        {
          id: 500,
          main: WeatherConditions.rain,
          description: 'light rain',
          icon: '10d',
        },
      ],
      wind_deg: 270,
      wind_speed: 5.1,
      pop: 0.2,
      wind_gust: 7.0,
      visibility: 10000,
      feels_like: 20 + i,
    })),
    lat: 0,
    lon: 0,
    timezone: 'UTC',
    timezone_offset: 0,
  };

  beforeEach(() => {
    mockUseGlobalContext.mockReturnValue({
      state: {
        weatherData: mockWeatherData,
        isLoading: false,
        location: 'London',
        coordinates: { lat: 0, lon: 0 },
        pageStyle: WeatherConditions.clear,
        clothesAdvice: {
          tempRangeName: TemperatureRangeNames.COLD,
          clothesStyle: ClothesStyle.basic,
          clothesDescription: '',
          clothesList: ['jacket', 'scarf'],
        },
      },
      fetchData: jest.fn(),
      handleSearch: jest.fn(),
    });
  });

  test('renders correctly when expanded', () => {
    render(<HourlyForecast expanded={true} />);
    const slider = screen.getByTestId('slider');
    expect(slider).toBeInTheDocument();
  });

  test('disables left arrow at initial position', () => {
    render(<HourlyForecast expanded={true} />);
    const leftArrow = screen.getByTestId('left-arrow');
    expect(leftArrow).toBeDisabled();
  });

  test('enables right arrow at initial position', () => {
    render(<HourlyForecast expanded={true} />);
    const rightArrow = screen.getByTestId('right-arrow');
    expect(rightArrow).not.toBeDisabled();
  });

  test('slides right and updates position', () => {
    render(<HourlyForecast expanded={true} />);
    const rightArrow = screen.getByTestId('right-arrow');

    fireEvent.click(rightArrow);

    const slider = screen.getByTestId('slider');
    expect(slider).toHaveStyle('transform: translateX(-55px)');
  });

  test('slides left and updates position', () => {
    render(<HourlyForecast expanded={true} />);
    const rightArrow = screen.getByTestId('right-arrow');
    const leftArrow = screen.getByTestId('left-arrow');

    fireEvent.click(rightArrow);
    fireEvent.click(leftArrow);

    const slider = screen.getByTestId('slider');
    expect(slider).toHaveStyle('transform: translateX(-0px)');
  });

  test('disables right arrow at max position', () => {
    render(<HourlyForecast expanded={true} />);
    const rightArrow = screen.getByTestId('right-arrow');

    for (let i = 0; i < 4; i++) {
      fireEvent.click(rightArrow);
    }

    expect(rightArrow).toBeDisabled();
  });

  test('matches snapshot', () => {
    const { container } = render(<HourlyForecast expanded={true} />);
    expect(container).toMatchSnapshot();
  });
});
