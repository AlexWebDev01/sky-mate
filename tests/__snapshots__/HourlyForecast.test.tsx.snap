// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HourlyForecast Component matches snapshot 1`] = `
<div>
  <div
    class="hourly-forecast"
  >
    <button
      class="left-arrow"
      data-testid="left-arrow"
      disabled=""
    >
      <svg
        data-testid="mocked-left-arrow"
      />
    </button>
    <div
      class="slider-container"
    >
      <div
        class="slider"
        data-testid="slider"
        id="slider"
        style="transform: translateX(-0px);"
      >
        <div
          class="day"
        >
          <div>
            21
            °
          </div>
          <div
            class="line"
          />
          <div>
            15
          </div>
        </div>
        <div
          class="day"
        >
          <div>
            22
            °
          </div>
          <div
            class="line"
          />
          <div>
            16
          </div>
        </div>
        <div
          class="day"
        >
          <div>
            23
            °
          </div>
          <div
            class="line"
          />
          <div>
            17
          </div>
        </div>
        <div
          class="day"
        >
          <div>
            24
            °
          </div>
          <div
            class="line"
          />
          <div>
            18
          </div>
        </div>
        <div
          class="day"
        >
          <div>
            25
            °
          </div>
          <div
            class="line"
          />
          <div>
            19
          </div>
        </div>
        <div
          class="day"
        >
          <div>
            26
            °
          </div>
          <div
            class="line"
          />
          <div>
            20
          </div>
        </div>
        <div
          class="day"
        >
          <div>
            27
            °
          </div>
          <div
            class="line"
          />
          <div>
            21
          </div>
        </div>
        <div
          class="day"
        >
          <div>
            28
            °
          </div>
          <div
            class="line"
          />
          <div>
            22
          </div>
        </div>
        <div
          class="day"
        >
          <div>
            29
            °
          </div>
          <div
            class="line"
          />
          <div>
            23
          </div>
        </div>
        <div
          class="day"
        >
          <div>
            30
            °
          </div>
          <div
            class="line"
          />
          <div>
            00
          </div>
        </div>
        <div
          class="day"
        >
          <div>
            31
            °
          </div>
          <div
            class="line"
          />
          <div>
            01
          </div>
        </div>
        <div
          class="day"
        >
          <div>
            32
            °
          </div>
          <div
            class="line"
          />
          <div>
            02
          </div>
        </div>
        <div
          class="day"
        >
          <div>
            33
            °
          </div>
          <div
            class="line"
          />
          <div>
            03
          </div>
        </div>
      </div>
    </div>
    <button
      class="right-arrow"
      data-testid="right-arrow"
    >
      <svg
        data-testid="mocked-right-arrow"
      />
    </button>
  </div>
</div>
`;
